community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java

```diff
@@ -302,77 +279,64 @@ private boolean nextBlock()
         // Read remaining data from previous property (if it was not read)
         while ( remainingBlocksToRead-- > 0 )
         {
+            // don't really care to retry since we discard the data...
             cursor.getLong();
         }
 
-        if ( cursor.getOffset() - offsetAtBeginning < PropertyStore.RECORD_SIZE )
+        int offset = cursor.getOffset();
+        if ( offset - offsetAtBeginning < PropertyStore.RECORD_SIZE )
         {
-            header = cursor.getLong();
+            try
+            {
+                do
+                {
+                    cursor.setOffset( offset );
+                    header = cursor.getLong();
+                }
+                while ( cursor.shouldRetry() );
+            }
+            catch ( IOException e )
+            {
+                throw new UnderlyingStorageException( e );
+            }
 
-            type = getPropertyType( header );
+            type = PropertyType.getPropertyType( header, true );
             if ( type != null )
             {
+                first = false;
                 keyId = (int) (header & KEY_BITMASK);
-
-                remainingBlocksToRead = type.calculateNumberOfBlocksUsed( header ) - 1;
-
-                originalHeader = header;
-                header = header >>> 28;
-
+                blocks = remainingBlocksToRead = type.calculateNumberOfBlocksUsed( header ) - 1;
                 return true;
             }
         }
 
-        cursor.close();
-        cursor = null;
-
-        return false;
+        if ( first )
+        {
+            throw new NotFoundException( "Property record with id " + currentRecordId + " not in use" );
+        }
+        else
+        {
+            cursor.close();
+            cursor = null;
+            return false;
+        }
     }
 
-    private long longFromIntAndMod( long base, long modifier )
+    private boolean valueIsInlined( long firstBlock )
     {
-        return modifier == 0 && base == IdGeneratorImpl.INTEGER_MINUS_ONE ? -1 : base | modifier;
+        // [][][][][   i,tttt][kkkk,kkkk][kkkk,kkkk][kkkk,kkkk]
+        return (firstBlock & 0x10000000L) > 0;
```

Comment:
Proposal is to have single utility that will do mask and shifting operations with descriptive names that both `PropertyBlock` and `StorePropertyCursor` will use.